### PR TITLE
Optimistic check for bucket presence without lock.

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli @geirst @toregge PR

### CODE REVIEW

File: storage/src/vespa/storage/bucketdb/btree_lockable_map.hpp [hunk 1]
```diff
@@ -179,6 +179,12 @@ void BTreeLockableMap<T>::acquireKey(const LockId& lid, std::unique_lock<std::mu
     }
 }
 
+template <typename T>
+bool BTreeLockableMap<T>::hasKey(const key_type& key) const {
+    auto readGuard = do_acquire_read_guard();
+    return _impl->find(key).valid();
```
Comment (toregge) [inline storage/src/vespa/storage/bucketdb/btree_lockable_map.hpp:185]: That find method is only supposed to be used by the writer and is not protected by the read guard.

**Review by toregge — CHANGES_REQUESTED** [review]

File: storage/src/vespa/storage/bucketdb/btree_lockable_map.hpp (hunk 1 above)
Comment (baldersheim) [reply storage/src/vespa/storage/bucketdb/btree_lockable_map.hpp]: That was what I suspected. new wiring added.

**Review by vekterli — COMMENTED** [review]
This change seems to introduce some subtle new interactions between mutating operations and operations that change the structure of the bucket tree. Should probably discuss this before deciding on whether we should merge.

**Comment by baldersheim** [comment]
Too risky...

**Pull request closed without merging**